global_helpers/panther_box_helpers.py

```diff
@@ -0,0 +1,124 @@
+BOX_JWT_PRIVATE_KEY = 'BOX_JWT_PRIVATE_KEY'
+BOX_JWT_PUB_KEY_ID = 'BOX_JWT_PUB_KEY_ID'
+BOX_ENTERPRISE_ID = 'BOX_ENTERPRISE_ID'
+BOX_JWT_KEY_PASSPHRASE = 'BOX_JWT_KEY_PASSPHRASE'  #nosec
+
+# Used to cache client connection to BOX API
+BOX_CLIENT = None
+
+# Cache credentials for 60 minutes at most
+#  (the default box access token lifetime)
+BOX_ACCESS_AGE = None
+MAX_BOX_ACCESS_AGE = 60
+
+# prevent INFO logs from going into the rules engine cloudtrail
+logging.getLogger('boxsdk').setLevel(logging.CRITICAL)
+
+
+class BadBoxLookup(Exception):
+    """Error returned when a box item lookup fails."""
+
+
+class BadSecretsLookup(Exception):
+    """Error returned when an AWS Secrets Manager lookup fails."""
+
+
+def is_box_sdk_enabled() -> bool:
+    return Client and JWTAuth
+
+
+def lookup_box_file(user_id: str, file_id: str) -> dict:
+    try:
+        client = get_box_client()
+        user = client.user(user_id=user_id)
+        file_info = client.as_user(user).file(file_id=file_id).get()
+        return file_info
+    except BoxAPIException as err:
+        raise BadBoxLookup('Exception looking up file info.') from err
+
+
+def lookup_box_folder(user_id: str, folder_id: str) -> dict:
+    try:
+        client = get_box_client()
+        user = client.user(user_id=user_id)
+        folder = client.as_user(user).folder(folder_id=folder_id).get()
+        return folder
+    except BoxAPIException as err:
+        raise BadBoxLookup('Exception looking up folder info.') from err
+
+
+def get_box_client() -> Client:
+    # pylint: disable=global-statement
+    global BOX_ACCESS_AGE
+    global BOX_CLIENT
+    if not Client or not JWTAuth:
+        raise Exception('Could not import necessary Box Library.')
+    if BOX_CLIENT is not None and BOX_ACCESS_AGE is not None and datetime.now(
+    ) - BOX_ACCESS_AGE < timedelta(minutes=MAX_BOX_ACCESS_AGE):
+        return BOX_CLIENT
+    response = boto3.client('secretsmanager').get_secret_value(
+        SecretId=BOX_API_ACCESS_NAME)
+    settings = build_jwt_settings(response)
+    BOX_CLIENT = Client(JWTAuth.from_settings_dictionary(settings))
+    BOX_ACCESS_AGE = datetime.now()
+    return BOX_CLIENT
+
+
+def build_jwt_settings(response: dict) -> dict:
+    data = None
+    if 'SecretString' in response:
+        data = response['SecretString']
+    else:
+        data = base64.b64decode(response['SecretBinary'])
+    # convert str from aws secrets mgr to json
+    data = json.loads(data)
+    # check that all necessary secrets are configured
+    if not all(key in data for key in [
+            BOX_CLIENT_ID, BOX_CLIENT_SECRET, BOX_JWT_PRIVATE_KEY,
+            BOX_JWT_PUB_KEY_ID, BOX_ENTERPRISE_ID, BOX_JWT_KEY_PASSPHRASE
+    ]):
+        raise BadSecretsLookup('Missing necessary secret(s): {}'.format([
+            BOX_CLIENT_ID, BOX_CLIENT_SECRET, BOX_JWT_PRIVATE_KEY,
+            BOX_JWT_PUB_KEY_ID, BOX_ENTERPRISE_ID, BOX_JWT_KEY_PASSPHRASE
+        ]))
```

Comment:
Might make sense to replace this with:
`raise BadSecretsLookup('Missing necessary secret(s): {}'.format(key))` 
As to alert the user that they are specifically missing the `key` field

Suppose `BOX_CLIENT_ID` is not defined, then this will raise:
`BadSecretsLookup - Missing necessary secrets(s): None, ....`
Which isn't as useful

Reply:
I think this does log the keys, not the values.  I didn't break it down into _which_ ones are missing - just listed them all.  I could add in a loop for logging just those that are missing, but I figured the user could also do the comparison. Do you prefer the former?

Reply:
This code may not be working the way you're thinking it does, example output here:
<img width="883" alt="example_output" src="https://user-images.githubusercontent.com/71197790/97924031-059f6c80-1d2d-11eb-961d-87a350b402a9.png">

I think there's value in telling the user which value is missing as opposed to a generic error message.

Reply:
I think it's generally better to provide specific information than a generic error message when possible

Reply:
Makes sense - I made the change, but as differences in sets rather than the loop to catch all possible missing keys rather than the first missing one.